[PyTorch][Edge] Add api to get bytecode version from runtime

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56948 [PyTorch][Edge] Add api to get bytecode version from runtime**



Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

Commit:
[PyTorch][Edge] Add api to get bytecode version from runtime

Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"


Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




1. Add api to get runtime bytecode version
2. This api is only available for mobile runtime

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




1. Add api to get runtime bytecode version
2. This api is only available for mobile runtime

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




1. Add api to get runtime bytecode version
2. This api is only available for mobile runtime

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

JacobSzwejbka commented on tools/build_variables.bzl:
```diff
@@ -373,6 +373,7 @@ libtorch_extra_sources = libtorch_core_jit_sources + [
     "torch/csrc/jit/mobile/module.cpp",
     "torch/csrc/jit/mobile/observer.cpp",
     "torch/csrc/jit/mobile/optim/sgd.cpp",
+    "torch/csrc/jit/mobile/runtime_bytecode_version.cpp",
```
Could you rename this to something like runtime_compatibility.cpp. I'm gonna do a similar thing to the model_bytecote.cpp when I add the model ops api. I think the seperation between model and runtime makes sense, but it also makes sense to me to put apis that will be used together as close together as possible

raziel commented on test/mobile/test_bytecode.py:
```diff
@@ -0,0 +1,10 @@
+class testVariousModelVersions(TestCase):
+    def test_get_runtime_bytecode_version(self):
+        runtime_bytecode_version = _get_runtime_bytecode_version()
+        assert(isinstance(runtime_bytecode_version, int))
```
maybe also >=0

iseeyuan replied:
int or int64_t?

cccclai replied:
int64_t is a macro we defined in cpp so we don't really have it in Python. `int` in Python should be sufficient to cover long type from cpp.

raziel commented on torch/csrc/jit/mobile/runtime_bytecode_version.cpp:
```diff
@@ -0,0 +1,14 @@
+
+namespace torch {
+namespace jit {
+namespace mobile {
```
Why does this have to live (i.e. be specific) to mobile? I understand bytecode is used for mobile, but kProducedBytecodeVersion is not mobile specific.

@iseeyuan

iseeyuan replied:
Good point. kProducedBytecodeVersion is not in mobile. The function name _get_runtime_bytecode_version is quite self explainable. kProducedBytecodeVersion is in caffe2::serialize where "caff2" is a legacy name. It's fine to me to just use torch::jit.

cccclai replied:
Resolved in the latest commit

raziel replied:
Also, does the code need to live in torch/csrc/jit/mobile/ as opposed to torch/csrc/jit/ ?

raziel reviewed (commented):
Thanks.

My main comment is about the implementation being mobile specific.

iseeyuan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1684,6 +1685,9 @@ void initJitScriptBindings(PyObject* module) {
         }
         return _load_for_mobile(in, optional_device);
       });
+  m.def("_get_runtime_bytecode_version", []() {
```
Python API for "lite interpreter runtime version" looks a little confusing to me. "runtime" is mostly used as a binary that can run on client. It's also the major purpose for this API. Adding a Python API means something from server side (because we don't assume Python be available on client). 

This API is telling the user which bytecode version the produced *model* would be,  not the actual version number deployed to the client. IMHO its good to expose to python the model API (provided a model file returns the version number inside that model), but not the runtime API, at least not with the name "runtime". Any thoughts @raziel ?

cccclai replied:
I was thinking this api can be use by client right before the model is loaded in client. They can do a quick bytecode compatibility check before the model is loaded. Maybe rename it to `_get_library_bytecode_version`?

iseeyuan replied:
My major concern is not on the naming. I don't think we should expose this API in python. If it's used in client, it's C++, not Python.

cccclai replied:
Remove it from Python in latest commit.

kimishpatel replied:
> My major concern is not on the naming. I don't think we should expose this API in python. If it's used in client, it's C++, not Python.

I agree with this with one caveat. If we ever have lite runtime deployed via python client. This might be the case in server but not usually the case in client. So I would still agree with @iseeyuan. If anything we might want to expose this via Java binding as well, if android apps want to convey the bytecode version that lite runtime can handle to model delivery.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM

cccclai commented:
Main change after the first round of review:
1. Rename runtime_bytecode_model file name to runtime_compatibility
2. Remove Python side api

raziel reviewed (changes requested):
I just want to settle whethere torch/csrc/jit/mobile/ is the right location for the API

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

cccclai commented:
> I just want to settle whethere `torch/csrc/jit/mobile/` is the right location for the API

Under `csrc/jit/mobile`, there are a few organized folders. https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit, including `api`, `backends`, `ir`, `passes`, `mobile` and etc. It looks to me that `mobile` is the appropriate folder to go.

I'm thinking that though it's not only specific to mobile (for example, can use in server to deploy model), it's mobile related. So it should be fine to leave it in `mobile`? @iseeyuan thought?

kimishpatel commented:
> > I just want to settle whethere `torch/csrc/jit/mobile/` is the right location for the API
> 
> Under `csrc/jit/mobile`, there are a few organized folders. https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit, including `api`, `backends`, `ir`, `passes`, `mobile` and etc. It looks to me that `mobile` is the appropriate folder to go.
> 
> I'm thinking that though it's not only specific to mobile (for example, can use in server to deploy model), it's mobile related. So it should be fine to leave it in `mobile`? @iseeyuan thought?

Given that this implementation is bytecode specific I would have thought that mobile is the right place? And move it when it becomes more generic? My 2 cents.

raziel reviewed (approved):
@cccclai @kimishpatel @iseeyuan 

Discussed this with @iseeyuan and we have a proposal for refactoring the code related to serialization from both producer and consumer side, which will create a better place to place this APIs (i.e. on the producer side).

@iseeyuan will write down the details (and create a task for us to follow up), but we agreed this refactoring can/should take place after the versioning changes since these take priority and there's not much cost to do that.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

iseeyuan commented:
> @cccclai @kimishpatel @iseeyuan
> 
> Discussed this with @iseeyuan and we have a proposal for refactoring the code related to serialization from both producer and consumer side, which will create a better place to place this APIs (i.e. on the producer side).
> 
> @iseeyuan will write down the details (and create a task for us to follow up), but we agreed this refactoring can/should take place after the versioning changes since these take priority and there's not much cost to do that.

Thanks for pinging me here! I commented in a wrong place. Pasted here:
File organization: it can be done in a separate PR later (tracking in T89985123)
* The serialization related codes (import.h/cpp, export_data.cpp, compatibility.h/cpp, etc) should be moved to torch/csrc/jit/serialization/mobile.
* torch/csrc/jit/mobile should be renamed to torch/csrc/jit/mobile_runtime to make this folder runtime specific.

As @raziel pointed out, it looks strange to have mobile model producer in jit/serialization, but model consumer + some data producer in jit/mobile. It makes more sense to move writeBytecode (which is directly used in jit/serialization/export_module), and all the serialization related code to jit/serialization/mobile. Since "mobile" is also specified there, people don't lose the context of "mobile specific".

kimishpatel commented:
I have missed this PR. Let me take a look, but if want to refactor mobile specific stuff, as @iseeyuan suggesetd, to serialization/mobile can I ask for this to wait-for-land before I land my delegate debug stuff? At least some of it. It will cause much pain to resolve all the merge issues.

kimishpatel commented:
> > @cccclai @kimishpatel @iseeyuan
> > Discussed this with @iseeyuan and we have a proposal for refactoring the code related to serialization from both producer and consumer side, which will create a better place to place this APIs (i.e. on the producer side).
> > @iseeyuan will write down the details (and create a task for us to follow up), but we agreed this refactoring can/should take place after the versioning changes since these take priority and there's not much cost to do that.
> 
> Thanks for pinging me here! I commented in a wrong place. Pasted here:
> File organization: it can be done in a separate PR later (tracking in T89985123)
> 
> * The serialization related codes (import.h/cpp, export_data.cpp, compatibility.h/cpp, etc) should be moved to torch/csrc/jit/serialization/mobile.
> * torch/csrc/jit/mobile should be renamed to torch/csrc/jit/mobile_runtime to make this folder runtime specific.
> 
> As @raziel pointed out, it looks strange to have mobile model producer in jit/serialization, but model consumer + some data producer in jit/mobile. It makes more sense to move writeBytecode (which is directly used in jit/serialization/export_module), and all the serialization related code to jit/serialization/mobile. Since "mobile" is also specified there, people don't lose the context of "mobile specific".

Makes sense. I hope I can block those PRs until I land the delegate debug stuff ;).

kimishpatel reviewed (commented):
Looks good! Thanks for this change. This will pave the way for us to get appropriate model delivered for the runtime, especially for people like me who dont update their apps that frequently :)

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

cccclai commented:
@kimishpatel Oops sorry I miss your massage. These prs are really important and need to be landed this coming week, so we can meet the branch cut day for next PyTorch release. (It's really soon...

kimishpatel commented:
> @kimishpatel Oops sorry I miss your massage. These prs are really important and need to be landed this coming week, so we can meet the branch cut day for next PyTorch release. (It's really soon...

Why? Why does refactor matter for next branch cut? And for that matter even this PR?

kimishpatel commented:
BTW I did not mean this PR for land blocking, but curious why does it matter for next branch cut? Is it tied to some feature we are releasing in next branch cut?

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode version from runtime"




Add api to get runtime bytecode version

## Test
Both `caffe2/test/cpp/jit/test_lite_interpreter.cpp` and `caffe2/test/mobile/test_bytecode.py` pass

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27987811/)!

Differential Revision: [D27987811](https://our.internmc.facebook.com/intern/diff/D27987811)

[ghstack-poisoned]

cccclai commented:
Chatted with @kimishpatel offline about how it relates to branch cut.
